### ISSUE TITLE
Pulling the keys out of a subset of a dictionary selected with Faker.PickRandom does not result in the expected output?

### ISSUE THREAD

**Issue by McPhale** [bchavez/Bogus#287]
### Version Information
| Software                 | Version(s)         |
| ------------------------ | ----------         |
| Bogus NuGet Package      | Bogus.28.4.4       |
| .NET Core?               | 2.1                |
| .NET Full Framework?     |                    |
| Windows OS?              | Windows 10         |
| Linux OS?                |                    |
| Visual Studio?           | Visual Studio 2017 |

### What locale are you using with Bogus?
En
### What's the problem?
Probably something I'm just not grocking, but I'm trying to pull the keys out of a randomized subset of a dictionary.
Instead of the expected output, it appears as if the keys are being randomly fetched as well.

Eg,

```cs
var dictionary = new Dictionary<string, string>();
dictionary.Add("key1", "val1");
dictionary.Add("key2", "val2");
dictionary.Add("key3", "val3");
dictionary.Add("key4", "val4");
dictionary.Add("key5", "val5");
dictionary.Add("key6", "val6");

var faker = new Faker();
var dictionarySubset = faker.PickRandom(dictionary, 3).Dump();
```
<img width="263" alt="AutobogusDictionaryOutput" src="https://user-images.githubusercontent.com/7131919/78067823-fc948a00-735c-11ea-9725-b14a016df882.PNG">

```cs
//(I would expect the output to be key2, key6, key5)
var subsetKeys = dictionarySubset.Select(s => s.Key).Dump();
```
Instead I end up with something like: 
<img width="168" alt="AutobogusKeyOutput" src="https://user-images.githubusercontent.com/7131919/78067850-05855b80-735d-11ea-9ba0-0603957db14c.png">

Can someone enlighten me on what's going on?

Thanks for the great library!

**Comment by bchavez** [comment]
Hi @McPhale ,

The TLDR is to call `.ToArray()` after your `.PickRandom()`:

```csharp
void Main()
{
   var dictionary = new Dictionary<string, string>();
   dictionary.Add("key1", "val1");
   dictionary.Add("key2", "val2");
   dictionary.Add("key3", "val3");
   dictionary.Add("key4", "val4");
   dictionary.Add("key5", "val5");
   dictionary.Add("key6", "val6");

   var faker = new Faker();
   //Call .ToArray() to evaluate the IEumerable Deferred Execution
   var dictionarySubset = faker.PickRandom(dictionary, 3).ToArray();
   dictionarySubset.Dump();

   var subsetKeys = dictionarySubset.Select(s => s.Key);
   subsetKeys.Dump();
   subsetKeys.Dump();
}
```
![image](https://user-images.githubusercontent.com/478118/78092752-6d4f9c80-7385-11ea-93a7-97bd2fe2e3d3.png)

-----

The long answer is `.PickRandom(IEnumerable, int count)` overload calls `.Random.Shuffle(IEnumerable).Take(count)` which creates a deferring execution.  In other words, every time you enumerate over `IEnumerable` you'll get a different subset of results unless you call `.ToArray()` to finalize enumeration:

```csharp
void Main()
{
   var dictionary = new Dictionary<string, string>();
   dictionary.Add("key1", "val1");
   dictionary.Add("key2", "val2");
   dictionary.Add("key3", "val3");
   dictionary.Add("key4", "val4");
   dictionary.Add("key5", "val5");
   dictionary.Add("key6", "val6");

   var faker = new Faker();
   //Call .ToArray() here to finalize the selection.
   var dictionarySubset = faker.PickRandom(dictionary, 3);
   dictionarySubset.Dump();

   //Or call .ToArray() here to finalize the selection.
   var subsetKeys = dictionarySubset.Select(s => s.Key);
   subsetKeys.Dump();
   subsetKeys.Dump();
   subsetKeys.Dump();
}
```
![image](https://user-images.githubusercontent.com/478118/78093356-e13e7480-7386-11ea-9275-dbd4bf0a8733.png)

-----

I'll have to think about changing this. The current behavior does seem a little weird at first glance. Especially when semantically `.PickRandom(IEnumerable)` and `.PickRandom(IEnumerable, int)` should be semantically the same. After looking at this it does seem, as you have discovered, there is a large divergence in behavior.

Thanks,
Brian

**Comment by McPhale** [comment]
o_o I would think I've been bitten enough times by deferred execution that I would be able to sniff it out by now! Thanks so much for the detailed explanation, that was very helpful.
